File: rust-2021/src/day06.rs
type Integer = usize;

pub fn read(inp: &str) -> [Integer; 7] {
    let mut v = [Integer::default(); 7];
    for word in inp.as_bytes().split(|&b| b == b',') {
        v[(word[0] - b'0') as usize] += 1;
    }
    v
}

fn solve(inp: &[Integer], n: usize) -> Integer {
    let mut counts = [0usize; 9];
    for (i, &c) in inp.iter().enumerate() {
        counts[i] = c;
    }
    for _ in 0..n {
        let new = counts[0];
        counts.rotate_left(1);
        counts[6] += new;
    }
    counts.into_iter().sum()
}

pub fn part1(inp: &[Integer]) -> Integer {
    solve(inp, 80)
}

pub fn part2(inp: &[Integer]) -> Integer {
    solve(inp, 256)
}

pub fn run() -> (String, String) {
    let inp = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/inputs/day06.input"));
    let data = read(inp);
    let a = part1(&data).to_string();
    let b = part2(&data).to_string();
    (a, b)
}

File: 2022/day06.py
import time
from functools import partial

from more_itertools import windowed

start = time.time()
with open("../inputs/2022/day06.input") as f:
    inp = f.read()


def solve(n: int, inp: str):
    return next(
        filter(lambda iwin: len(set(iwin[1])) == n, enumerate(windowed(inp, n), n))
    )[0]


part1 = partial(solve, 4)
part2 = partial(solve, 14)


print(
    f"Day 06     {part1(inp):<14} {part2(inp):<14} {(time.time() - start)*1e3:>11.2f}"
)
